PDF print preview shows blank preview with UI-side compositing

nmoucht opened the pull request:
#### 94dbfb8994c360197362ec90dac20b5e85082ed6
<pre>
PDF print preview shows blank preview with UI-side compositing
<a href="https://bugs.webkit.org/show_bug.cgi?id=248785">https://bugs.webkit.org/show_bug.cgi?id=248785</a>
&lt;rdar://102826227&gt;

Reviewed by Simon Fraser.

Create local buffer in the WebProcess so that the PDF print preview
generation code can use the CGContext to draw the PDF.

* Source/WebCore/platform/graphics/RenderingMode.h:
* Source/WebKit/Shared/ImageOptions.h:
* Source/WebKit/Shared/WebImage.cpp:
(WebKit::WebImage::create):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createImageBuffer const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::drawRectToImage):

Canonical link: <a href="https://commits.webkit.org/257667@main">https://commits.webkit.org/257667@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d4a20f507a65b3cc82942b2f83d2978d60d8f23c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99587 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8762 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32685 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108949 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169186 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103583 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9349 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86068 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92060 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106866 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105362 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90588 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [💥 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34033 "An unexpected error occured. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Updated gtk dependencies; Pull request contains relevant changes; Skipped layout-tests; layout-tests (exception)") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21945 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2601 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23459 "Passed tests") | | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2535 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45851 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8680 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42932 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4395 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2700 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -5793,11 +5793,13 @@ void WebPage::drawRectToImage(FrameIdentifier frameID, const PrintInfo& printInf
+        RetainPtr<PDFDocument> pdfDocument = pdfDocumentForPrintingFrame(coreFrame);
 #else
         ASSERT(coreFrame->document()->printing());
+        bool pdfDocument = false;
```
Please don't use a different type on different branches of the #ifdef. Refactor the code to avoid this.

nmoucht replied:
Changed this code to always use a local buffer for print preview.

smfr commented on Source/WebKit/Shared/ImageOptions.h:
```diff
@@ -30,6 +30,7 @@ namespace WebKit {
 
 enum ImageOptions {
     ImageOptionsShareable = 1 << 0,
+    ImageOptionsLocalPDFSnapshot = 1 << 1,
```
This should be something more generic, like "Use local buffer".

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/graphics/RenderingMode.h:
```diff
@@ -34,7 +34,8 @@ enum class RenderingPurpose : uint8_t {
     ShareableSnapshot,
-    MediaPainting
+    MediaPainting,
+    LocalBufferForSnapshot
```
We should have some consistency between these options:
```
Snapshot,
ShareableSnapshot
LocalBufferForSnapshot
```

`LocalBufferForSnapshot` makes a Shareable (with the UI-process) buffer. So let's call the new value `ShareableLocalSnapshot`, and put it before `MediaPainting`.

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/Shared/ImageOptions.h:
```diff
@@ -30,6 +30,7 @@ namespace WebKit {
 
 enum ImageOptions {
     ImageOptionsShareable = 1 << 0,
+    ImageOptionsLocalBuffer = 1 << 1,
```
`ImageOptionsLocal`, but with a comment saying what "local" means.

nmoucht replied:
Fixed.

nmoucht committed:
PDF print preview shows blank preview with UI-side compositing
https://bugs.webkit.org/show_bug.cgi?id=248785
<rdar://102826227>

Reviewed by Simon Fraser.

Create local buffer in the WebProcess so that the PDF print preview
generation code can use the CGContext to draw the PDF.

* Source/WebCore/platform/graphics/RenderingMode.h:
* Source/WebKit/Shared/ImageOptions.h:
* Source/WebKit/Shared/WebImage.cpp:
(WebKit::WebImage::create):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::createImageBuffer const):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::drawRectToImage):

Canonical link: https://commits.webkit.org/257667@main
